Make Unix terminal interface bindings domain-safe

OlivierNicole opened the pull request:
For unclear reasons (maybe to save stack space), Unix.tcgetattr and Unix.tcsetattr both modify a global variable of type `struct termios`. To make the bindings domain-safe, this replaces the global variable with stack variables.

I would need to test that the new bindings are correct, but I'm not experienced in serial programming, so I'm submitting this draft PR in case someone has advice on the test to perform. Maybe try to compile and use an ncurses package?

OlivierNicole committed:
Make Unix terminal interface bindings domain-safe

Up to now, Unix.tcgetattr and Unix.tcsetattr both modify a global
variable of type `struct termios`.
To make the bindings domain-safe, this replaces the global variable with
stack-allocated variables.

xavierleroy commented:
> For unclear reasons (maybe to save stack space), Unix.tcgetattr and Unix.tcsetattr both modify a global variable of type struct termios. 

I guess the reason was to put pointers (not offsets) in the `terminal_io_descr` table.  Nothing deep.

> To make the bindings domain-safe, this replaces the global variable with stack variables.

Excellent idea, thanks!

> I would need to test that the new bindings are correct, but I'm not experienced in serial programming, so I'm submitting this draft PR in case someone has advice on the test to perform. Maybe try to compile and use an ncurses package?

In general, these are bindings to ncurse-like C libraries, which perform terminal manipulations themselves in C.  What about an OPAM-wide grep for tc[sg]etattr ?

You can get some confidence by comparing (visually) the OCaml results with the output of `stty -a`, see below for a sample session (obtained with your branch).  But please don't try to turn this into a proper unit test: during CI, there's no guarantee that a pseudo-terminal is available, and many things concerning pseudo-terminals and serial links are system-specific, so this would be (another) CI nightmare.

```
$ runtime/ocamlrun ./ocaml -I ./stdlib -I ./otherlibs/unix
OCaml version 5.1.0+dev1-2022-06-09
Enter #help;; for help.

# #load "unix.cma";;
# Unix.(tcgetattr stdin);;
- : Unix.terminal_io =
{Unix.c_ignbrk = false; c_brkint = false; c_ignpar = false; c_parmrk = false;
 c_inpck = false; c_istrip = false; c_inlcr = false; c_igncr = false;
 c_icrnl = true; c_ixon = true; c_ixoff = false; c_opost = true;
 c_obaud = 38400; c_ibaud = 38400; c_csize = 8; c_cstopb = 1; c_cread = true;
 c_parenb = false; c_parodd = false; c_hupcl = false; c_clocal = false;
 c_isig = true; c_icanon = true; c_noflsh = false; c_echo = true;
 c_echoe = true; c_echok = true; c_echonl = false; c_vintr = '\003';
 c_vquit = '\028'; c_verase = '\127'; c_vkill = '\021'; c_veof = '\004';
 c_veol = '\000'; c_vmin = 1; c_vtime = 0; c_vstart = '\017';
 c_vstop = '\019'}
# Sys.command "stty -a";;
speed 38400 baud; rows 24; columns 80; line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>;
eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R;
werase = ^W; lnext = ^V; discard = ^O; min = 1; time = 0;
-parenb -parodd -cmspar cs8 -hupcl -cstopb cread -clocal -crtscts
-ignbrk -brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl ixon -ixoff
-iuclc -ixany -imaxbel iutf8
opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt
echoctl echoke -flusho -extproc
- : int = 0
Unix.(tcsetattr stdin TCSANOW { (tcgetattr stdin) with c_ibaud = 9600 });;
- : unit = ()
# Sys.command "stty -a";;
speed 9600 baud; rows 24; columns 80; line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>;
eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R;
werase = ^W; lnext = ^V; discard = ^O; min = 1; time = 0;
-parenb -parodd -cmspar cs8 -hupcl -cstopb cread -clocal -crtscts
-ignbrk -brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl ixon -ixoff
-iuclc -ixany -imaxbel iutf8
opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt
echoctl echoke -flusho -extproc
- : int = 0
```

OlivierNicole commented:
Will do that, thanks.

OlivierNicole commented on otherlibs/unix/termios.c:
```diff
@@ -314,12 +313,11 @@ static int when_flag_table[] = {
 
 CAMLprim value caml_unix_tcsetattr(value fd, value when, value arg)
 {
-  if (tcgetattr(Int_val(fd), &terminal_status) == -1)
-    caml_uerror("tcsetattr", Nothing);
-  decode_terminal_status(&Field(arg, 0));
```
I did not quite understand why `tcsetattr` calls `tcgetattr`, for me it does not change the result so I removed it, but I just want to mention this in case I'm missing something.

OlivierNicole replied:
Well removing this does change the semantics. As of now, `tcsetattr` updates the fields of its `terminal_io` argument. I don't quite see when this can be useful. But if existing code relies on this, maybe we don't want to change it.

xavierleroy replied:
In general, `struct termios` contains additional, OS-specific fields and bits beyond the standard ones that are mapped to the `Unix.terminal_io` OCaml type.  It's better not to change these additional fields.

OlivierNicole replied:
Understood, I put the call to `tcgetattr` back.

shindere commented:
Xavier Leroy (2022/08/04 01:32 -0700):
> You can get some confidence by comparing (visually) the OCaml results
> with the output of `stty -a`, see below for a sample session (obtained
> with your branch).  But please don't try to turn this into a proper
> unit test: during CI, there's no guarantee that a pseudo-terminal is
> available, and many things concerning pseudo-terminals and serial
> links are system-specific, so this would be (another) CI nightmare.

It might be possible to introduce a guard test that makes sure the test
is run on a real terminal. In that case, the test would be skipped on CI
but run when the testsuite is run manually on a developer's system.
Would that sound a sane approach to you, @xavierleroy?

OlivierNicole commented:
The new code looks clearly domain-safe to me, and I have performed Xavier's test and checked that setting and getting characteristics works as expected with all setting types (Bool, Enum, Char and Speed). Couldn't we decide that the new functions are very likely to be correct and that a dedicated test is not necessary?

xavierleroy committed:
Unix termios interface: use POSIX type tcflag_t

Since POSIX.1, the 4 "flag" members of `struct termios` have type `tcflag_t`.
Hence, use `tcflag_t` instead of `int` to access them.

xavierleroy commented:
> Couldn't we decide that the new functions are very likely to be correct and that a dedicated test is not necessary?

Yes, we can decide that no test is necessary.  The new functions should not be any more buggy than the old ones :-)

I took the liberty to push an extra commit to this branch that uses the POSIX-specified type `tcflag_t` to access the flag fields of the struct, instead of `int` as before.  Under Linux, `tcflag_t` is `unsigned int`, so it makes no difference there.

I'll run a round of CI precheck to check that it doesn't break anything.

xavierleroy commented:
Oh, also: could you please remove the Draft status on this PR?  Otherwise it cannot be merged.

xavierleroy reviewed (approved):
CI seems happy.  Please add a Changes entry and remove the Draft flag.

OlivierNicole committed:
Add Changes entry

OlivierNicole commented:
Done.
